### ISSUE TITLE
A few missing API methods in the Criteria API

### ISSUE THREAD

**Issue by archiecobbs** [eclipse-ee4j/jpa-api#190]
The following methods are inexplicably missing from the Criteria API:

* `CriteriaBuilder.treat(From, Class)`
* `Subquery.correlate(From)`

As a result, programmers have to use a sibling method with a wider type, causing them to have to add unchecked casts to compensate, etc.

**Comment by gavinking** [comment]
@archiecobbs I'm not sure I understand. Typically when creating a criteria query you will have an instance of `Root` or `Join` to work with, since there aren't any important methods which return `From`.

Do you have a specific motivating code example or scenario which would help justify this change?

**Comment by archiecobbs** [comment]
Hi @gavinking,

I discovered this omission while developing [QueryStream](https://github.com/querystream/querystream/), which is a "wrapper" library around the Criteria API that tries to make coding type-safe Criteria queries simpler. So yes it is probably accurate to say these methods would not be regularly used by normal people, but they do represent "holes" in the API in the sense that support for every other case is provided and these holes cause problems for such a "wrapper" library.

For the first case, check out [`PathStream.cast()`](https://querystream.github.io/querystream/site/apidocs/index.html?org/dellroad/querystream/jpa/PathStream.html). The method returns the invoked object for method chaining and therefore has return type `PathStream`. Now consider [`FromStream`](https://querystream.github.io/querystream/site/apidocs/index.html?org/dellroad/querystream/jpa/FromStream.html) which is a subtype of `PathStream`. Ideally, it should also have an overriding method `FromStream.cast()` that returns `FromStream`. But this is impossible to implement, because there's no version of `Criteria.treat()` that accepts a `From`.

Similarly for `Subquery.correlate()`, checkout the [`QueryStream.Builder`](https://querystream.github.io/querystream/site/apidocs/index.html?org/dellroad/querystream/jpa/QueryStream.Builder.html) API and the `substream()` methods it provides. The others can be implemented in a single line, while the `substream(From)` variant has to explicitly try all the other cases, and fails if an actual `From` is ever encountered (even though that may rarely happen):
```java
        @SuppressWarnings("unchecked")
        public <X, Y> FromStream<Y, ? extends From<X, Y>> substream(From<X, Y> from) {
            if (from == null)
                throw new IllegalArgumentException("null join");
            if (from instanceof Root)
                return (FromStream<Y, ? extends From<X, Y>>)this.substream((Root<Y>)from);
            if (from instanceof SetJoin)
                return this.substream((SetJoin<X, Y>)from);
            if (from instanceof ListJoin)
                return this.substream((ListJoin<X, Y>)from);
            if (from instanceof MapJoin)
                return this.substream((MapJoin<X, ?, Y>)from);
            if (from instanceof CollectionJoin)
                return this.substream((CollectionJoin<X, Y>)from);
            if (from instanceof Join)
                return this.substream((Join<X, Y>)from);
            throw new UnsupportedOperationException("substream() from " + from.getClass().getName());
        }
```

Thanks.

P.S. I'm curious what you think about this library in general.

**Comment by gavinking** [comment]
@archiecobbs most of that `switch` is unnecessary, I  believe, because `Join` is a super type of all the other join types. So you only need two cases in the `switch`, I believe.

**Comment by archiecobbs** [comment]
Probably true. Regardless, that code is an ugly and incomplete workaround for the real problem. I look forward to discarding it when this ticket gets fixed :)

**Comment by gavinking** [comment]
I mean, I gotta be honest: you're not really selling me on this. If the only real payoff here is to eliminate the need for a single one-line function:

```java
static <X,Y> Join<X,Y> correlate(Subquery<?> subquery, From<X,Y> from) {
    return from instanceof Join ? subquery.correlate((Join<X,Y>) from) : subquery.correlate((Root<X,Y>) from);
}
```

which is only needed in _specialized library code_, then I would say let's not do this. It just doesn't look to me like a massive burden to maintain that function.

**Comment by archiecobbs** [comment]
You seem to be asserting that every `From` instance is always actually either a `Root` or a `Join` for every JPA implementation that exists or may someday exist.

Even if that might happen to be true today, it's not safe to just assume that it always will be... unless you write that into the specification.

In any case, my point is not that no workaround exists, but rather that, in principle, a properly designed API should not be forcing workarounds like this in the first place.

That's why this bug is an API specification bug, not a "it doesn't work" type of bug.

**Comment by archiecobbs** [comment]
Bleh. I ran into this again and (it took me 10 minutes to track down this bug, I knew it existed somewhere...).

To restate, at best there is an inconsistency. We have all of these methods:
```
<X,​T,​E extends T> CollectionJoin<X,​E> treat​(CollectionJoin<X,​T> join, Class<E> type)
<X,​T,​V extends T> Join<X,​V> 	      treat​(Join<X,​T> join, Class<V> type)
<X,​T,​E extends T> ListJoin<X,​E>       treat​(ListJoin<X,​T> join, Class<E> type)
<X,​K,​T,​V extends T> MapJoin<X,​K,​V>    treat​(MapJoin<X,​K,​T> join, Class<V> type)
<X,​T extends X> Path<T>               treat​(Path<X> path, Class<T> type)
<X,​T extends X> Root<T>               treat​(Root<X> root, Class<T> type)
<X,​T,​E extends T> SetJoin<X,​E>        treat​(SetJoin<X,​T> join, Class<E> type)
```
but not:
```
<X,​T, V extends T> From<X, V> 	      treat​(From<X, T> path, Class<V> type)
```
Why not?? That just seems obviously inconsistent and weird.

Rhetorical question: Why do all those methods exist in the first place? Why not just replace them all with this and require everyone to perform unchecked casts?
```
Selection<?> 	treat​(Selection<?> path, Class<V> type)
```
Answer: because unchecked casts are ugly and well-designed API's minimize the need for them.

Plus this would be so trivially easy to fix....

I don't understand the resistance.

**Comment by gavinking** [comment]
> I don't understand the resistance.

The resistance is that if we do every single thing that one person asks for once, then our APIs grow unboundedly with a forest of operations that aren't generally useful. There absolutely has to be at least some friction here.

Still, that fact that you still remember this eight years later is a positive signal. So I have put something together.

**Comment by archiecobbs** [comment]
> > I don't understand the resistance.
> 
> The resistance is that if we do every single thing that one person asks for once, then our APIs grow unboundedly with a forest of operations that aren't generally useful. There absolutely has to be at least some friction here.

Yes I'm not a fan of unnecessary bloat either. But I think this is a little different... there is a `treat()` method, and because of the way Java typing works, you need a separate override for each of the possible input types. Having an override for every input type except `From` is kind of like if we had `String.valueOf()` overrides for everything except `float`, i.e., that would just seem oddly incomplete.

> Still, that fact that you still remember this eight years later is a positive signal. So I have put something together.

Thanks for taking a look at it!

**Comment by gavinking** [comment]
Done.
